service/main.go

```diff
@@ -58,25 +58,29 @@ type indexOptions struct {
 
 // service struct handler for all necessary fiels for indexing
 type service struct {
-	indexer     indexer
-	provider    provider
-	driver      driver
-	hasEnd      bool
-	fromHeight  int
-	toHeight    int
-	retries     int64
-	concurrency int64
-	reqInterval time.Duration
+	indexer          indexer
+	fallbackIndexer  indexer
+	provider         provider
+	fallbackProvider provider
+	driver           driver
+	hasEnd           bool
+	fromHeight       int
+	toHeight         int
+	retries          int64
+	concurrency      int64
+	reqInterval      time.Duration
 }
 
-func newService(retries, concurrency int64, reqInterval time.Duration, provider provider, driver driver, indexer indexer, options *indexOptions) (*service, error) {
+func newService(retries, concurrency int64, reqInterval time.Duration, provider, fallbackProvider provider, driver driver, indexer, fallbackIndexer indexer, options *indexOptions) (*service, error) {
```

Comment:
those are a lot of params, why don't you better make this a config struct? and add some validation method within the struct to avoid all the checking ifs in this function